src/server/http_server.rs

```
@@ -0,0 +1,138 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use futures::future::{ok, FutureResult};
+use futures::sync::oneshot::{Receiver, Sender};
+use futures::{self, Future};
+use hyper::service::service_fn;
+use hyper::{self, Body, Method, Request, Response, Server, StatusCode};
+use tokio_threadpool::{Builder, ThreadPool};
+
+use std::net::SocketAddr;
+use std::str::FromStr;
+
+use super::Result;
+use util::metrics::dump;
+
+pub struct HttpServer {
+    thread_pool: ThreadPool,
+    tx: Sender<()>,
+    rx: Option<Receiver<()>>,
+    addr: Option<SocketAddr>,
+}
+
+impl HttpServer {
+    pub fn new(http_thread_pool_size: usize) -> Self {
+        let thread_pool = Builder::new()
+            .pool_size(http_thread_pool_size)
+            .name_prefix("http-server-")
+            .after_start(|| {
+                info!("HTTP server started");
+            })
+            .before_stop(|| {
+                info!("stopping HTTP server");
+            })
+            .build();
+        let (tx, rx) = futures::sync::oneshot::channel::<()>();
+        HttpServer {
+            thread_pool,
+            tx,
+            rx: Some(rx),
+            addr: None,
+        }
+    }
+
+    pub fn start(&mut self, http_addr: String) -> Result<()> {
+        let addr = SocketAddr::from_str(&http_addr)?;
+
+        // TODO: support TLS for HTTP server.
+        let builder = Server::try_bind(&addr)?;
+
+        // Create an HTTP service.
+        let service = |req: Request<Body>| -> FutureResult<Response<Body>, hyper::Error> {
+            let mut response = Response::new(Body::empty());
+
+            match (req.method(), req.uri().path()) {
+                (&Method::GET, "/metrics") => match dump() {
+                    Err(_) => {
+                        *response.status_mut() = StatusCode::INTERNAL_SERVER_ERROR;
```

don't ignore the error here, you can return the error message as the body.